### PR TITLE
Add SuperCollapsedBlock

### PR DESCRIPTION
**Pull request by KevinBoulongne**
Depends on #1706
Depends on #1683
Depends on https://github.com/Infomaniak/android-core/pull/125

### CODE REVIEW

**Review by FabianDevel — CHANGES_REQUESTED** [review]

File: app/src/main/java/com/infomaniak/mail/ui/main/thread/ThreadAdapter.kt [hunk 1]
```diff
@@ -105,50 +107,87 @@ class ThreadAdapter(
...
                     failedLoadingErrorMessage.isVisible = true
-                    if (isExpandedMap[message.uid] == true) onExpandedMessageLoaded(message.uid)
+                    if (isExpandedMap[item.uid] == true) onExpandedMessageLoaded(item.uid)
                 }
                 NotifyType.ONLY_REBIND_CALENDAR_ATTENDANCE -> {
-                    val attendees = message.latestCalendarEventResponse?.calendarEvent?.attendees ?: emptyList()
+                    val attendees = item.latestCalendarEventResponse?.calendarEvent?.attendees ?: emptyList()
                     holder.binding.calendarEvent.onlyUpdateAttendance(attendees)
                 }
             }
```
[the remark this answers is not in the document]
Comment (KevinBoulongne) [reply app/src/main/java/com/infomaniak/mail/ui/main/thread/ThreadAdapter.kt:159]: Done in another PR: https://github.com/Infomaniak/android-kMail/pull/1707

File: app/src/main/java/com/infomaniak/mail/ui/main/thread/ThreadViewModel.kt [hunk 2]
```diff
@@ -95,29 +118,139 @@ class ThreadViewModel @Inject constructor(
...
+        }
+
+        return items to messagesToFetch
+    }
+
+    private fun computeFirstIndexAfterBlock(thread: Thread, list: RealmResults<Message>): Int {
+        val firstDefaultIndex = list.count() - 2
+        val firstUnreadIndex = if (thread.unseenMessagesCount > 0) list.indexOfFirstOrNull { !it.isSeen } else null
+        val notNullFirstUnreadIndex = firstUnreadIndex ?: firstDefaultIndex
+        return minOf(notNullFirstUnreadIndex, firstDefaultIndex)
```
Comment (FabianDevel) [inline app/src/main/java/com/infomaniak/mail/ui/main/thread/ThreadViewModel.kt:215]:
```suggestion
        val notNullFirstUnreadIndex = firstUnreadIndex ?: firstDefaultIndex

        return minOf(notNullFirstUnreadIndex, firstDefaultIndex)
```

File: app/src/main/java/com/infomaniak/mail/ui/main/thread/ThreadViewModel.kt [hunk 3]
```diff
@@ -95,29 +118,139 @@ class ThreadViewModel @Inject constructor(
...
+                    previousBlock.contains(message.uid) && isStillInBlock -> { // All Messages already in block
+                        superCollapsedBlock!!.add(message.uid)
+                    }
+                    !previousBlock.contains(message.uid) && isStillInBlock -> { // First Message not in block
+                        isStillInBlock = false
+                        items += SuperCollapsedBlock(superCollapsedBlock!!)
+                        addMessage(message.apply { shouldHideDivider = true })
+                    }
+                    else -> { // All following Messages
+                        addMessage(message)
```
Comment (FabianDevel) [inline app/src/main/java/com/infomaniak/mail/ui/main/thread/ThreadViewModel.kt:187]: I don't really get this logic, I'll need explanations

**Review by FabianDevel — CHANGES_REQUESTED** [review]

File: app/src/main/java/com/infomaniak/mail/ui/main/thread/ThreadViewModel.kt [hunk 4]
```diff
@@ -95,29 +118,139 @@ class ThreadViewModel @Inject constructor(
...
+        messages: RealmResults<Message>,
+        threadUid: String,
+    ): Pair<ThreadAdapterItems, MessagesWithoutHeavyData> {
+
+        val items = mutableListOf<Any>()
+        val messagesToFetch = mutableListOf<Message>()
+        val thread = messages.firstOrNull()?.threads?.firstOrNull { it.uid == threadUid } ?: return items to messagesToFetch
+        val firstIndexAfterBlock = computeFirstIndexAfterBlock(thread, messages)
+        shouldDisplaySuperCollapsedBlock = shouldSuperCollapsedBlockBeDisplayed(messages.count(), firstIndexAfterBlock)
+        val shouldCreateSuperCollapsedBlock = shouldSuperCollapsedBlockBeCreated()
```
Comment (LunarX) [inline app/src/main/java/com/infomaniak/mail/ui/main/thread/ThreadViewModel.kt:137]: The method `shouldSuperCollapsedBlockBeCreated` has a side effect by instantiating an empty mutable map inside a function that should only return a boolean

[the remark this answers is not in the document]
Comment (KevinBoulongne) [reply app/src/main/java/com/infomaniak/mail/ui/main/thread/ThreadViewModel.kt:137]: Nope, it's used in the following inner functions.

**Review by FabianDevel — APPROVED** [review]

File: app/src/main/java/com/infomaniak/mail/ui/main/thread/calendar/AttendeesBottomSheetDialog.kt [hunk 5]
```diff
@@ -48,9 +48,8 @@ class AttendeesBottomSheetDialog : BottomSheetDialogFragment() {
         root.title = context.getString(R.string.attendeesListTitle, attendees.count())
         attendeeRecyclerView.adapter = AttendeesAdapter(attendees)
 
-        val dividerItemDecorator = DividerItemDecorator(
-            InsetDrawable(UiUtils.dividerDrawable(context), 16.toPx(), 0, 16.toPx(), 0)
-        )
-        attendeeRecyclerView.addItemDecoration(dividerItemDecorator)
+        val margin = context.resources.getDimensionPixelSize(R.dimen.dividerHorizontalPadding).toPx()
```
Comment (tevincent) [inline app/src/main/java/com/infomaniak/mail/ui/main/thread/calendar/AttendeesBottomSheetDialog.kt:51]: `context.resources.getDimensionPixelSize(R.dimen.dividerHorizontalPadding)` already return somthing in pixel. We don't need to convert this value in pixel again.

File: app/src/main/java/com/infomaniak/mail/ui/newMessage/RecipientFieldView.kt [hunk 6]
```diff
@@ -360,10 +357,11 @@ class RecipientFieldView @JvmOverloads constructor(
         onToggleEverythingCallback: ((isCollapsed: Boolean) -> Unit)? = null,
     ) {
 
-        val dividerItemDecorator = DividerItemDecorator(InsetDrawable(dividerDrawable(context), 16.toPx(), 0, 16.toPx(), 0))
+        val margin = context.resources.getDimensionPixelSize(R.dimen.dividerHorizontalPadding).toPx()
```
Comment (tevincent) [inline app/src/main/java/com/infomaniak/mail/ui/newMessage/RecipientFieldView.kt:360]: Also here.

**Review by tevincent — CHANGES_REQUESTED** [review]

**Pull request merged**